### PR TITLE
Add mermaid export

### PR DESCRIPTION
**Pull request by lkuchenb**
* Adds export function that returns a string that can be parsed by mermaid
* Classes can be either shown with or without the properties

### CODE REVIEW

**Review by KerstenBreuer — COMMENTED** [review]
Looks great at a first glance. Would you maybe like to queue this PR after this one #25? Then you could already add a cli.

**Comment by lkuchenb** [comment]
Sure 👍

**Comment by lkuchenb** [comment]
@KerstenBreuer ready for another round, CLI command added

File: tests/test_cli.py [hunk 1]
```diff
@@ -168,3 +169,46 @@ def test_check_datapack_not_complies():
     result = runner.invoke(cli, ["check-datapack", str(datapack)])
     assert result.exit_code == exit_codes.DATAPACK_SPEC_ERROR != 0
     assert "DataPackSpecError" in result.stderr
+
+
+@pytest.mark.parametrize(
+    "props,example_suffix", [([], "_wo_props.mm.txt"), (["-p"], "_w_props.mm.txt")]
```
Comment (KerstenBreuer) [inline tests/test_cli.py:175]: Also check with the long arguments i.e. here `--with-properties` to ensure that the interface is stable. Might also be worth to align with the solutions chosen above to make the test look more uniform.

File: tests/test_export_mermaid.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+from schemapack import load_schemapack
+from schemapack.export.mermaid import export_mermaid
+from tests.fixtures.examples import EXAMPLES_DIR, VALID_SCHEMAPACK_PATHS
+
+
+@pytest.mark.parametrize(
+    "with_properties, file_suffix",
+    [(True, "_w_props.mm.txt"), (False, "_wo_props.mm.txt")],
+)
```
Comment (KerstenBreuer) [inline tests/test_export_mermaid.py:29]: Rather take the same approach as for the schemapack and datapack examples:

- Automatically ready all erd diagram examples in the tests.fixture.examples to generate an ERD_PATHS dict
- use a naming convention so that, from the name of the erd example, the corresponding schemapack as well as the parameters (w/o props) can be deduced

Like that you can simply add more files to the example directory and they will be automatically picked up by the tests.

**Review by KerstenBreuer — CHANGES_REQUESTED** [review]
Nice and solid solution. Just the embedding into the rest of the code base might be tweaked.

File: src/schemapack/_internals/cli/main.py [hunk 3]
```diff
@@ -117,3 +119,53 @@ def check_datapack(
...
 
     print_final_success("The provided document complies with the specs of a datapack.")
+
+
+@cli.command()
+def export_mermaid(
+    *,
+    schemapack: Annotated[
+        Path,
+        typer.Argument(
```
Comment (KerstenBreuer) [inline src/schemapack/_internals/cli/main.py:129]: In the other commands, paths to schemapacks and datapacks are always given by command line options not arguments. We should probably make this consistent.

File: examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+array[string] comments "opt"
+}
+Product }|--|{ ProductCategory : "Product.categories"
+ProductCategory {
+string name "req"
+}
+
+ShippingEvent {
+string datetime "req"
+}
```
Comment (KerstenBreuer) [inline examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt:33]: Maybe use some nice indentation:
```suggestion
erDiagram
    Customer {
        string name "req"
        object attributes "req"
        boolean company "req"
    }
    Customer ||--o{ Address : "Customer.addresses"
    Customer ||--o{ Order : "Customer.orders"
    Address {
        string street "req"
        string city "req"
        integer zip_code "req"
        enum country "req"
    }
    
    Order {
        string datetime "req"
    }
    Order }o--|{ Product : "Order.products"
    Order ||--o| ShippingEvent : "Order.shipping_event"
    Product {
        string name "req"
        number price "req"
        array[string] comments "opt"
    }
    Product }|--|{ ProductCategory : "Product.categories"
    ProductCategory {
        string name "req"
    }
    
    ShippingEvent {
        string datetime "req"
    }
```

File: examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt [hunk 5]
```diff
@@ -0,0 +1,33 @@
+erDiagram
```
Comment (KerstenBreuer) [inline examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt:1]: Maybe group relations at the top. Then add classes with content below.

File: tests/test_export_mermaid.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+def test_export_mermaid(with_properties: bool, file_suffix: str):
```
Comment (KerstenBreuer) [inline tests/test_export_mermaid.py:30]: Maybe also add edge cases, like:
- having no content properties
- having properties that are of type object or list

Comment (KerstenBreuer) [inline tests/test_export_mermaid.py:30]: How do we deal with additionalProperties being set to true? Maybe add a test for the expected behavior as well.

**Review by KerstenBreuer — COMMENTED** [review]
Found some additional small suggestions

**Comment by KerstenBreuer** [comment]
Could you maybe also bump the version? I forgot to do that in the CLI PR.

File: src/schemapack/_internals/cli/main.py [hunk 7]
```diff
@@ -117,3 +119,53 @@ def check_datapack(
...
+            "--with-properties",
+            "-p",
+            help="Include properties in the output.",
+        ),
+    ] = False,
+    force: Annotated[
+        bool,
+        typer.Option(
+            "--force",
+            "-f",
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply src/schemapack/_internals/cli/main.py:154]: I would prefer to keep this unchanged, since '-f / --force' is a pretty well established flag for descructive behaviours, most popularly overwriting outputs.

File: src/schemapack/_internals/cli/main.py [hunk 8]
```diff
@@ -117,3 +119,53 @@ def check_datapack(
...
+    ],
+    output: Annotated[
+        Path,
+        typer.Argument(
+            help="Provide the path to the mermaid output file to produce.",
+            file_okay=True,
+            dir_okay=False,
+            writable=True,
+        ),
+    ],
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply src/schemapack/_internals/cli/main.py]: Output is now written to stdout equivalently to the other cli commands 👍

File: src/schemapack/_internals/cli/io.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+from contextlib import contextmanager
+from pathlib import Path
+
+import typer
+
+from schemapack._internals.cli import exit_codes
+from schemapack._internals.cli.printing import print_final_failure
+
+
+@contextmanager
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply src/schemapack/_internals/cli/io.py:25]: No longer needed with stdout.

File: examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt (hunk 4 above)
Comment (lkuchenb) [reply examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt]: Content properties are indented now.

File: examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt (hunk 5 above)
Comment (lkuchenb) [reply examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt:1]: Do you mean [as in the mermaid examples](https://mermaid.js.org/syntax/entityRelationshipDiagram.html#attributes) (per class) or do you mean globally all relations first, then all classes? I'm not sure what would be the advantage, unless I'm missing any would prefer to keep the current styl.e

File: src/schemapack/_internals/export/mermaid.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply src/schemapack/_internals/export/mermaid.py]: I've restructured the module structure accordingly. Just as a background, some application make a distinction between "save"/"load" and "export"/"import", where the former is for the application native format and the latter is for non-native format(s) which might be lossy or only importable or exportable. Thus I thought this might actually provide a proper distinction.

File: src/schemapack/_internals/cli/main.py [hunk 11]
```diff
@@ -117,3 +119,53 @@ def check_datapack(
...
+    force: Annotated[
+        bool,
+        typer.Option(
+            "--force",
+            "-f",
+            help="Overwrite the output file if it exists.",
+        ),
+    ] = False,
+):
+    """Export a schemapack to a markdown file diagram."""
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply src/schemapack/_internals/cli/main.py]: Fixed.

File: src/schemapack/export/mermaid.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply src/schemapack/export/mermaid.py]: Fixed.

File: tests/test_export_mermaid.py (hunk 6 above)
Comment (lkuchenb) [reply tests/test_export_mermaid.py:29]: Good point, I've added an asterisk row for that case and added it to the test schema as well. 
Screenshot 2024-04-16 at 22 28 50

Comment (lkuchenb) [reply tests/test_export_mermaid.py:29]: array was covered in the tests, i've added the object case, too 👍

File: tests/test_cli.py (hunk 1 above)
Comment (lkuchenb) [reply tests/test_cli.py]: Fixed in 5f3daaf90ba97fcf6840536e7be9f17c6483f83d

File: tests/test_export_mermaid.py (hunk 2 above)
Comment (lkuchenb) [reply tests/test_export_mermaid.py]: Fixed in 5cfe4bcd123a257120794f27337ffa765e894b3f

**Comment by lkuchenb** [comment]
@KerstenBreuer ready for another round

File: examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt (hunk 5 above)
Comment (KerstenBreuer) [reply examples/mermaid/comprehensive_cardinalities_and_types_w_props.mm.txt]: I meant first putting all these lines:

```
Customer ||--o{ Address : "Customer.addresses"
Customer ||--o{ Order : "Customer.orders"
...
```

And then all the class definitions such as:
```
Customer {
string name "req"
object attributes "req"
boolean company "req"
}
```

But this is really personal taste and doesn't really matter that much.

I see that in the example you provided they interleave the two. So maybe this is the common pratice.

File: src/schemapack/_internals/cli/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply src/schemapack/_internals/cli/main.py]: Alright, no objections.

File: src/schemapack/_internals/export/mermaid.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply src/schemapack/_internals/export/mermaid.py]: OK, I was not aware of that. Makes also sense. At least once we have multiple export targets.

File: examples/erd/comprehensive_cardinalities_and_types_w_props.mm.txt [hunk 13]
```diff
@@ -0,0 +1,40 @@
+erDiagram
```
Comment (KerstenBreuer) [inline examples/erd/comprehensive_cardinalities_and_types_w_props.mm.txt:1]: Looks very readable now, thanks.

**Review by KerstenBreuer — APPROVED** [review]
Very nice, tested it and works flawelessly.

**Review by KerstenBreuer — APPROVED** [review]
I fixed the merge conflict and approved again. I hope that was OK?

**Comment by lkuchenb** [comment]
> I fixed the merge conflict and approved again. I hope that was OK?

Very much appreciated ;-) Thanks
